sysl2/sysl/deps.go

```diff
@@ -0,0 +1,264 @@
+						r = append(r, item)
+					}
+					return true
+				}
+				if !hasPattern("human", app.GetAttrs()) {
+					r = append(r, item)
+				}
+				return true
+			})
+		}
+		return true
+	})
+
+	return MakeStrSet(r...)
+}
+
+func walkPassthrough(excludes, passthroughs StrSet, dep *AppDependency, integrations *DependencySet, module *sysl.Module) {
+	target := dep.Target
+	targetName := target.Name
+	targetEndpoint := target.Endpoint
+	excludedApps := MakeStrSet(targetName).Intersection(excludes)
+	undeterminedEps := MakeStrSet(targetEndpoint).Intersection(MakeStrSet(".. * <- *", "*"))
+	//Add to integration array since all dependencies are determined.
+	if len(excludedApps) == 0 && len(undeterminedEps) == 0 {
+		integrations.Deps[dep.String()] = dep
+	}
+
+	// find the next outbound dep
+	if passthroughs.Contains(targetName) {
+		cs := NewCallStatementSlice()
+		cs.CollectCallStatements(module.GetApps()[targetName].GetEndpoints()[targetEndpoint].GetStmt())
+		for _, stmt := range cs.GetSlice() {
+			call := stmt.GetStmt().(*sysl.Statement_Call).Call
+			nextAppName := strings.Join(call.GetTarget().GetPart(), " :: ")
+			nextEpName := call.GetEndpoint()
+			next := MakeAppElement(nextAppName, nextEpName)
+			nextDep := MakeAppDependency(target, next, stmt)
+			walkPassthrough(excludes, passthroughs, nextDep, integrations, module)
+		}
+	}
+}
+
+func (ds *DependencySet) FindIntegrations(apps, excludes, passthroughs StrSet, module *sysl.Module) *DependencySet {
+	integrations := NewDependencySet()
+	outboundDeps := NewDependencySet()
+	lenPassthroughs := len(passthroughs)
+	for _, dep := range ds.Deps {
+		appNames := dep.extractAppNames()
+		endpoints := dep.extractEndpoints()
+		isSubsection := isSub(appNames, apps)
+		isSelfSubsection := isSub(MakeStrSet(dep.Self.Name), apps)
+		isTargetSubsection := isSub(MakeStrSet(dep.Target.Name), passthroughs)
+		interExcludes := appNames.Intersection(excludes)
+		interEndpoints := endpoints.Intersection(MakeStrSet(".. * <- *", "*"))
+		lenInterExcludes := len(interExcludes)
+		lenInterEndpoints := len(interEndpoints)
+		if isSubsection && lenInterExcludes == 0 && lenInterEndpoints == 0 {
+			integrations.Deps[dep.String()] = dep
+		}
+		// collect outbound dependencies
+		if lenPassthroughs > 0 &&
+			((isSubsection || (isSelfSubsection && isTargetSubsection)) && lenInterExcludes == 0 && lenInterEndpoints == 0) {
+			outboundDeps.Deps[dep.String()] = dep
+		}
+	}
+	if lenPassthroughs > 0 {
+		for _, dep := range outboundDeps.Deps {
+			walkPassthrough(excludes, passthroughs, dep, integrations, module)
+		}
+	}
+
+	return integrations
+}
+
+func (ds *DependencySet) CollectAppDependencies(module *sysl.Module) {
+	for appname, app := range module.GetApps() {
+		for epname, endpoint := range app.GetEndpoints() {
+			ds.collectStatementDependencies(endpoint.GetStmt(), appname, epname)
+		}
+	}
+}
+
+func (ds *DependencySet) collectStatementDependencies(stmts []*sysl.Statement, appname, epname string) {
```

Comment:
@JackLeeHal `collectStatementDependencies` and `CollectCallStatements` look same. 
can you just create a common function like python has `yield_call_statements`?
This duplicate code worries me that we are looping multiple times on the same thing.

Reply:
Add a map into DependencySet to store call statements for each group by "appname:epname", then delete CollectCallStatements method.